### PR TITLE
[#522] Add a wait + synchronous `Workflow.create`

### LINKED ISSUE #522: Add `wait` method to `Workflow`
**Background**

To help perform live tests of workflow templates, we would like to be able to write the following in test cases:

```python
wflo_template = WorkflowTemplate(...)
wflo_template.apply(...)
wflo = wflo_template.run_as_worklow(...)
wflo.wait()
# Check post-conditions on workflow
```

**Proposal**

Add a `wait` method to `Workflow` that returns only after the workflow has been completed, or a timeout has been reached.

### PR DESCRIPTION
**Pull request by flaviuvadan**
**Pull Request Checklist**
- [x] Fixes #522
- [ ] Tests added
- [x] Documentation/examples added
- [x] [Good commit messages](https://cbea.ms/git-commit/) and/or PR title

**Description of PR**

See #522 for the feature request!

### CODE REVIEW

File: src/hera/workflows/workflow.py [hunk 1]
```diff
@@ -330,13 +332,44 @@ def to_yaml(self, *args, **kwargs) -> str:
...
         assert self.workflows_service, "workflow service not initialized"
         assert self.namespace, "workflow namespace not defined"
-        return self.workflows_service.create_workflow(
+        if async_:
+            return self.workflows_service.create_workflow(
+                WorkflowCreateRequest(workflow=self.build()), namespace=self.namespace
+            )
+        return self.wait(poll_interval=poll_interval)
+
+    def wait(self, poll_interval: int = 5) -> TWorkflow:
```
Comment (elliotgunton) [inline src/hera/workflows/workflow.py:355]: What do we think about a timeout parameter as well? We could then also allow a cleanup callback hook if it times outs and e.g. send a `WorkflowTerminateRequest` to kill the workflow

Comment (flaviuvadan) [reply src/hera/workflows/workflow.py:355]: Sounds like an amazing idea! What do you think about doing that in a follow up PR? We should also discuss how that will play with `active_deadline_seconds`: https://argoproj.github.io/argo-workflows/fields/#fields_3

Comment (elliotgunton) [reply src/hera/workflows/workflow.py:355]: Sgtm! Won't be a breaking change so fine by me

**Review by elliotgunton — APPROVED** [review]

File: src/hera/workflows/workflow.py [hunk 2]
```diff
@@ -330,13 +332,44 @@ def to_yaml(self, *args, **kwargs) -> str:
...
+        """Waits for the Workflow to complete execution.
+
+        Parameters
+        ----------
+        poll_interval: int = 5
+            The interval in seconds to poll the workflow status.
+        """
+        wf = self.workflows_service.create_workflow(
```
[the remark this answers is not in the document]
Comment (elliotgunton) [reply src/hera/workflows/workflow.py:363]: Yep that's a bug

[the remark this answers is not in the document]
Comment (flaviuvadan) [reply src/hera/workflows/workflow.py]: I aimed for a design in which Hera does not have to track the name of a workflow based on users' use of `generated_name=True`. Not performing a create inside the wait implies users have to call create first, which is ok, or call create with the wait flag (replaced async_ with wait). I made the adjustment and now there's a get inside the wait!

File: src/hera/workflows/workflow.py [hunk 3]
```diff
@@ -330,13 +332,57 @@ def to_yaml(self, *args, **kwargs) -> str:
...
+        """
+        assert self.workflows_service is not None, "workflow service not initialized"
+        assert self.namespace is not None, "workflow namespace not defined"
+        assert self.name is not None, "workflow name not defined"
+
+        wf = self.workflows_service.get_workflow(self.name, namespace=self.namespace)
+        assert wf.status is not None, f"workflow status not defined for workflow {wf.metadata.name}"
+        assert wf.status.phase is not None, f"workflow phase not defined for workflow status {wf.status}"
+        status = WorkflowStatus.from_argo_status(wf.status.phase)
+        assert wf.metadata.name is not None, f"workflow name not defined for workflow {self.name}"
```
Comment (elliotgunton) [inline src/hera/workflows/workflow.py:376]: Think it makes more sense to have the name assert first as the status assert uses it. Added a whitespace for readability
```suggestion
        wf = self.workflows_service.get_workflow(self.name, namespace=self.namespace)
        assert wf.metadata.name is not None, f"workflow name not defined for workflow {self.name}"

        assert wf.status is not None, f"workflow status not defined for workflow {wf.metadata.name}"
        assert wf.status.phase is not None, f"workflow phase not defined for workflow status {wf.status}"
        status = WorkflowStatus.from_argo_status(wf.status.phase)
```

**Review by samj1912 — APPROVED** [review]

**Pull request merged**